ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -284,11 +284,8 @@ runWith RunNodeArgs{..} LowLevelRunNodeArgs{..} =
                 , ChainDB.cdbVolatileDbValidation  = ValidateAll
                 }
 
-      (_, chainDB) <- allocate registry
-        (\_ -> openChainDB
-          registry inFuture cfg initLedger
-          llrnChainDbArgsDefaults customiseChainDbArgs')
-        ChainDB.closeDB
```

Comment:
How do we ensure now that the ChainDB is closed? I know we now do:

```haskell
    _ <- lift $ allocate (Args.cdbRegistry args) (\_ -> return $ chainDB) API.closeDB
```

but is it a problem if the exception now occurs before that line?

Reply:
If the exception occurs before that line the open states will be deallocated (and all their resources) as well as any threads that have been registered in the upper registry (the cache and background tasks threads) so everything would be clean. Also the chaindb itself won't be yet stored in any resulting state (as it is still not returned) and therefore it is not a problem that the TVar itself is not modified to `ChainDbClosed` as it will not be used. Furthermore, no followers or iterators could have been created so there's nothing else to free.

Reply:
To be super precise, we could register `env` in the temporal registry and couple it with a closing function (specifically `closeDB` not to be confused with `API.closeDB`) and free the resources manually.